boot/bootstate20.go

```diff
@@ -37,20 +41,49 @@ func newBootState20(typ snap.Type) bootState {
 	}
 }
 
+// newEarlyBootState20 is like newBootState20, but it initializes the
+// bootState with the provided bootloader and/or modeenv - mainly to be used
+// with the initramfs logic where we load the modeenv and bootloader slightly
+// differently than normal.
+func newEarlyBootState20(typ snap.Type) earlyBootState {
+	switch typ {
+	case snap.TypeBase:
+		bs := &bootState20Base{}
+		bs.earlyBoot = true
+		return bs
+	case snap.TypeKernel:
+		bs := &bootState20Kernel{earlyBoot: true}
+		return bs
+	case snap.TypeSnapd:
+		bs := &bootState20Snapd{}
+		bs.earlyBoot = true
+		return bs
+	default:
+		panic(fmt.Sprintf("cannot make a bootState20 for snap type %q", typ))
+	}
+}
+
 //
 // modeenv methods
 //
 
 type bootState20Modeenv struct {
 	modeenv *Modeenv
+
+	// this just controls where to try to find the modeenv
+	earlyBoot bool
```

Comment:
these should carry something more concrete than a flag, if the flag made huge behavior differences then there should be different structs, it seems to control only where are things so maybe these structs should carry where are things